src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/CustomAuthenticatorFactory.php

```diff
@@ -44,27 +46,55 @@ public function getKey(): string
     public function addConfiguration(NodeDefinition $builder)
     {
         $builder
-            ->info('An array of service ids for all of your "authenticators"')
-            ->requiresAtLeastOneElement()
-            ->prototype('scalar')->end();
+            ->fixXmlConfig('service')
+            ->beforeNormalization()
+                ->ifTrue(function ($v) { return is_string($v) || (is_array($v) && !isset($v['services']) && !isset($v['entry_point'])); })
+                ->then(function ($v) {
+                    return ['services' => (array) $v];
+                })
+            ->end()
+            ->children()
+                ->arrayNode('services')
+                    ->info('An array of service ids for all of your "authenticators"')
+                    ->requiresAtLeastOneElement()
+                    ->prototype('scalar')->end()
+                ->end()
+                ->scalarNode('entry_point')->defaultNull()->end()
+            ->end();
 
         // get the parent array node builder ("firewalls") from inside the children builder
         $factoryRootNode = $builder->end()->end();
         $factoryRootNode
             ->fixXmlConfig('custom_authenticator')
-            ->validate()
-                ->ifTrue(function ($v) { return isset($v['custom_authenticators']) && empty($v['custom_authenticators']); })
-                ->then(function ($v) {
-                    unset($v['custom_authenticators']);
-
-                    return $v;
-                })
-            ->end()
         ;
     }
 
     public function createAuthenticator(ContainerBuilder $container, string $firewallName, array $config, string $userProviderId): array
     {
-        return $config;
+        return $config['services'];
+    }
+
+    public function registerEntryPoint(ContainerBuilder $container, string $id, array $config): ?string
+    {
+        if (isset($config['entry_point'])) {
+            return $config['entry_point'];
+        }
+
+        $entryPoints = [];
+        foreach ($config['services'] as $authenticatorId) {
+            if (class_exists($authenticatorId) && is_subclass_of($authenticatorId, AuthenticationEntryPointInterface::class)) {
```

Comment:
Sadly, we can't do `$container->findDefinition($authenticatorId)->getClass()` here, as the `ContainerBuilder` is scoped for this extension.

Is there some magic trick we can do here to also allow non-class service IDs?